### DISCUSSION TITLE
Supporting search for all versions of OneDrive

### DISCUSSION THREAD

**Discussion by spencerwooo** [spencerwooo/onedrive-vercel-index#295] [Announcements]
<img width="1598" alt="Screenshot 2022-01-21 at 22 17 34" src="https://user-images.githubusercontent.com/32114380/150542285-2dd888ca-bfd5-45af-ba27-587f0c38aa55.png">

- Search supported with Microsoft Graph API ~~is expected to land tomorrow~~ **has landed!**
- Development branch and PR is at https://github.com/spencerwooo/onedrive-vercel-index/pull/283

Caveats:

- Limited support as the search provided by Microsoft Graph API is not ideal.
- Extremely limited support for CJK search.

...so, your mileage may vary. But it's better than nothing, right? 🙃

**Comment by spencerwooo** [comment]
Native search support has landed! Use `Ctrl` or `⌘ + K` to open the search box, and `ESC` to close.

~~**OneDrive business users - your instance may not work, as the path of the file is not returned from OneDrive API.** 😢 See issue #299.~~

**Edit: We now have full support (I hope) for OneDrive for Business and SharePoint users! (See screen recording below).**

Besides caveats we mentioned before:

- Search results still include your protected directories
- This means your **file names** (whether inside the protected folder or not) are exposed
- Still, visitors still need to enter passwords if they want to access the protected file contents

**Comment by aozaki-kuro** [comment]
好家伙，所以目前使用 OneDrive for Business 的话是不能建立索引的吗…
偏偏大量文件挂载又只能扔 Business，只能说不愧是微软了😂
晚些我自己也弄一下看看是不是普遍存在的问题……

**Reply by spencerwooo** [reply]
主要是：https://github.com/spencerwooo/onedrive-vercel-index/issues/299

并且我现在我自己 E5 开发者订阅的 OneDrive for Business 目前搜不出来东西，所以也没办法修……

**Reply by aozaki-kuro** [reply]
我这里部署了发现也是同样的情况，用无修改的原项目代码测试也是拉取不了任何搜索结果😂 Redis 也已经删掉重新授权过了。
目前用的是 A1 订阅，手里还有个 E5 但是内容没转过去，暂时没法测…

**Reply by code004Accepted** [reply]
看来不是我一个人的问题😅我也是E5，搜不出来东西

**Comment by spencerwooo** [comment]
https://user-images.githubusercontent.com/32114380/150788337-e8e2db9c-ccfa-4297-acb6-ec7f2f8220ad.mov

> Left: OneDrive for Business (E5 Developer Subscription). Right: normal OneDrive international version.

**🥳  We now support search also for OneDrive for Business / SharePoint!**

- You may notice some differences, that's because for OneDrive for Business, we need an additional query to the OneDrive API to request the `path` of the file.
- So, OneDrive for Business users, you would see a loading text when your search results first load.
- We also leverage `useSWR` here to cache the search results so that minimal data is queried each session.

Enjoy. ❤️

**Reply by spencerwooo** [reply]
If nothing gets returned no matter what you search - **that's on OneDrive's f*king stupid search API**, not on me.

**Reply by aozaki-kuro** [reply]
跟进部署了，果然还是不行，应该就是 1drv 那个残废搜索 API 的锅了😅
Edit: 可以搜索文件夹，不能搜索文件……

**Reply by spencerwooo** [reply]
等……说不定哪天 OneDrive 一抽风，就索引好了呢（反正今天就是早上我还搜不出来东西，下午突然就能搜了）

**Reply by zkeq** [reply]
> 等……说不定哪天 OneDrive 一抽风，就索引好了呢（反正今天就是早上我还搜不出来东西，下午突然就能搜了）

可以成功搜索了,但是发现一个bug:

当文件夹含有`#`号时,不能成功跳转,跳转链接为 `#`  但是正常的url路径应为`%23`

测试地址: 到 `https://onedrive.zkeq.xyz` 搜索 `test-url.mp4` 点击第一个,出现此错误....

删除文件夹名中的`#`号 再测试即恢复正常

**Reply by spencerwooo** [reply]
> 当文件夹含有 `#` 时，不能成功跳转，跳转链接为 `#` 但是正常的 url 路径应为 %23

OneDrive API 返回的路径没有正确做 URL encode，最新 commit 应该修复了这个问题。

**Comment by aozaki-kuro** [comment]
跟进一下，目前已经部署，但是我这里的情况是能搜出带数字的文件夹，但是连 README.md 这样简单的文件一个都搜不出来。
在 1drv 页面里是需要点击 “查看更多搜索结果” 才能看到相应结果…不知道这里是不是还有影响
